linkerd/app/core/src/control.rs

```diff
@@ -54,16 +58,46 @@ impl Config {
         B::Error: Into<Error> + Send + Sync,
         L: Clone + Param<tls::client::Config> + Send + 'static,
     {
-        let backoff = {
+        let connect_backoff = {
             let backoff = self.connect.backoff;
             move |_| Ok(backoff.stream())
         };
+
+        // When a DNS resolution fails, log the error and use the TTL, if there
+        // is one, to drive re-resolution attempts.
+        let resolve_backoff = {
+            let backoff = self.connect.backoff;
+            move |error: Error| {
+                warn!(%error, "Failed to resolve control-plane component");
+                if let Some(e) = error.downcast_ref::<dns::ResolveError>() {
+                    if let dns::ResolveErrorKind::NoRecordsFound {
+                        negative_ttl: Some(ttl_secs),
+                        ..
+                    } = e.kind()
+                    {
+                        let ttl = time::Duration::from_secs(*ttl_secs as u64);
+                        let stream = stream! {
+                            loop {
+                                time::sleep(ttl).await;
+                                yield;
+                            }
+                        };
```

Comment:
this is equivalent to taking a [`tokio::time::Interval`](https://docs.rs/tokio/1.2.0/tokio/time/fn.interval.html) and wrapping it as a `Stream`, either via [`tokio-stream`](https://docs.rs/tokio-stream/0.1.3/tokio_stream/wrappers/struct.IntervalStream.html) or our `IntoStream` utility in `linkerd-channel`. might be a little nicer to just use `Interval` (`Interval` is also a little more efficient but this use-case isn't really performance-critical).